front: handle timetable filter with exceptions

SharglutDev opened the pull request:
close #11472

SharglutDev committed:
front: handle timetable filter with exceptions

Signed-off-by: SharglutDev <p.filimon75@gmail.com>

SharglutDev committed:
front: adapt timetable filter tests for exceptions

Signed-off-by: SharglutDev <p.filimon75@gmail.com>

Math-R commented on front/src/modules/trainschedule/components/Timetable/useFilterTimetableItems.ts:
```diff
@@ -72,35 +90,69 @@ const useFilterTimetableItems = (
             series = '',
             subseries = '',
           } = timetableItem.rollingStock?.metadata || {};
+          rollingStockMetadata.push(detail, family, reference, series, subseries);
```
Is it on purpose that you choose to push those value flat in an array ?
Would it not be simplier to make an array of object ?
If some of the values ​​are `''` by default, don't you risk losing the sense of what's what?

Math-R replied:
And made some errors because of the potential non clear position of each value in the array ?

SharglutDev replied:
> If some of the values ​​are `''` by default, don't you risk losing the sense of what's what?

We already dealt with empty strings before so I don't think it might cause some issues.

> Is it on purpose that you choose to push those value flat in an array ? Would it not be simplier to make an array of object ?

I don't think so, the filter input wants to filter regardless if it comes from the detail, family etc so I think we can consider all of them as part of the same "characteristic" and put them in a flat array (there might be some duplicates in it but as it's a small array, I didn't bother remove the duplicates)

Math-R replied:
Alright, not a big fan of losing those information, but it's not blocking for me.

Math-R commented on front/src/modules/trainschedule/components/Timetable/utils.ts:
```diff
@@ -1,18 +1,29 @@
   if (searchString) {
+    let hasMatchingExceptions = false;
+    if (isPacedTrainWithDetails(item)) {
+      hasMatchingExceptions = item.exceptions.some(
+        (exception) =>
+          exception.train_name?.value.toLowerCase().includes(searchString.toLowerCase()) ||
+          exception.labels?.value.some((label) =>
+            label.toLowerCase().includes(searchString.toLowerCase())
```
Nit: Maybe you can extract those two test in specific and testable function,
one testing by the train_name the other by the label ?

SharglutDev replied:
Hmm I didn't want to loop two times on the exception for that since we look both time on information nested in it. Would it be ok for you to leave it that way ?

Math-R replied:
Yes sure that was nit :)

Math-R commented:
Also, outside the scope of this PR, would it be possible to consider a slightly more in-depth refactoring of this filter function?
It's increasingly looking like a large, unreadable monolith.
I think it might be interesting to separate all the different predicates into unit-testable functions. And then apply them in a useMemo.
By doing this, we could also organize the execution order of each predicate after monitoring which ones reject the most items?
What do you think?
@clarani @emersion @kmer2016

Maymanaf reviewed (approved):
LGTM for e2e tests ✅

RomainValls commented on front/src/modules/trainschedule/components/Timetable/useFilterTimetableItems.ts:
```diff
@@ -72,35 +90,69 @@ const useFilterTimetableItems = (
+                .map((exception) => extractTagCode(exception.speed_limit_tag!.value))
+            : [];
+          const allTags = uniq([itemTag, ...exceptionTags]);
+          // Si aucun tag de l'item ou de ses exceptions n'est dans selectedTags, on filtre
```
comment should be in english 👍

SharglutDev replied:
Fixed :)

RomainValls reviewed (commented):
I tried filtering with the "name, label" input : if the service name is "A" and the occurences "A1", "A2","A3" , ... , the filter works when searching for "A" but not "A1", "A2", ... . 
Not sure if it's wanted or not 😄

SharglutDev commented:
> I tried filtering with the "name, label" input : if the service name is "A" and the occurences "A1", "A2","A3" , ... , the filter works when searching for "A" but not "A1", "A2", ... . Not sure if it's wanted or not 😄

Hmm yes, good catch, that's more like a miss from the filter changes in the paced train issues from last PI.
Here we take count of occurrence name in the filter only if they have a name change group in exception.
Should I fix that here @maelysLeratRosso or do we open a bug ?

maelysLeratRosso commented:
> > I tried filtering with the "name, label" input : if the service name is "A" and the occurences "A1", "A2","A3" , ... , the filter works when searching for "A" but not "A1", "A2", ... . Not sure if it's wanted or not 😄
> 
> Hmm yes, good catch, that's more like a miss from the filter changes in the paced train issues from last PI. Here we take count of occurrence name in the filter only if they have a name change group in exception. Should I fix that here @maelysLeratRosso or do we open a bug ?

I think it would be good to handle it here as well if possible ?

SharglutDev committed:
front: fix occurrence names not handled in timetable filter

Signed-off-by: SharglutDev <p.filimon75@gmail.com>

SharglutDev commented:
Pushed a new commit to handle occurrence names :)

RomainValls reviewed (approved):
LGTM, tested 👍

Math-R reviewed (approved):
lgtm !

SharglutDev committed:
front: fix flaky test one paced train item click

Probably since the dynamic size of the timetable column since the new
scenario page was merge, the click on the paced train item sometimes
occurs on the non-clickable zone of the item.

To fix it :
- remove an unnecessary margin-right on the left part of the item. It
  was shrinking the clickable zone and was not on the mockup specs.
- on tests side, catch the item by the name element and not its parent.
  This ensures that the click will always work.

Signed-off-by: SharglutDev <p.filimon75@gmail.com>

SharglutDev committed:
front: fix flaky test for label filter

The test expected to have only one paced train displayed when it should
have been 2. Change the load state so it properly waits for the filter
to be done before expecting the train count.

Signed-off-by: SharglutDev <p.filimon75@gmail.com>

SharglutDev commented:
Waiting for https://github.com/OpenRailAssociation/osrd/pull/12255 that should fix the last failing tests
